### PR TITLE
[0.39] Feat/dydx perpetual connector

### PR DESCRIPTION
**Pull request by trkoneill18**
**Before submitting this PR, please make sure**:

- [ ] Your code builds clean without any errors or warnings
- [ ] You are using approved title ("feat/", "fix/", "docs/", "refactor/")

**A description of the changes proposed in the pull request**:

These commits add the dydx perpetual connector to the connector/derivative folder.

Note that in order to use this connector, you will need to generate a number of credentials, including a private stark key from dydx and api key credentials. These can be obtained by using https://docs.dydx.exchange/?python#derive-starkkey and https://docs.dydx.exchange/?python#recover-default-api-credentials, respectively.

One feature of this connector is that it will set the leverage to the maximum allowable leverage for dydx if the specified leverage is greater than this maximum.

### CODE REVIEW

**Comment by PtrckM** [comment]
- new connector color should be yellow, missing on supported connectors https://github.com/trkoneill18/hummingbot/blob/feat/dydx-perpetual-connector/README.md
image
image

- getting error when doing `balance` command
image
```
2021-04-10 06:56:36,552 - 41844 - hummingbot.core.utils.async_utils - ERROR - Unhandled error in background task: [<class 'decimal.DivisionUndefined'>]
Traceback (most recent call last):
  File "/home/ubuntu/dydx-perpetual-connector/hummingbot/core/utils/async_utils.py", line 9, in safe_wrapper
    return await c
  File "/home/ubuntu/dydx-perpetual-connector/hummingbot/client/command/balance_command.py", line 99, in show_balances
    self._notify(f"\n  Total: {RateOracle.global_token_symbol} {smart_round(df[total_col_name].sum())}    "
decimal.InvalidOperation: [<class 'decimal.DivisionUndefined'>]
```

**Comment by rxlxrxsx** [comment]
Prompt for leverage still shows Binance Perpetual, leverage should also be 25x maximum leverage based on their [website](https://help.dydx.exchange/en/articles/4800587-l2-perpetual-contract-specs) except for LINK-USD which has a maximum of 10x.

Screen Shot 2021-04-11 at 1 44 58 PM

**Comment by rxlxrxsx** [comment]
With position management set to profit_taking, once an order is filled to open a position the opposing order is not cancelled immediately, not honoring order_refresh_time.

Example:

At `11:04:58` my sell order gets filled, with order_refresh_time set to 60, the opposing buy order should be cancelled at `11:05:38` however it only got cancelled at `11:14:39`.

Screen Shot 2021-04-12 at 11 05 58 AM

```
2021-04-12 11:04:37,001 - 3701 - hummingbot.strategy.perpetual_market_making.perpetual_market_making - INFO - (ETH-USD) Creating 1 LIMIT bid orders at (Size, Price): ['0.01 ETH, 2117.3 USD'] to OPEN position.
2021-04-12 11:04:37,002 - 3701 - hummingbot.strategy.perpetual_market_making.perpetual_market_making - INFO - (ETH-USD) Creating 1  LIMIT ask orders at (Size, Price): ['0.01 ETH, 2138.9 USD'] to OPEN position.
2021-04-12 11:04:38,297 - 3701 - hummingbot.connector.derivative.dydx_perpetual.dydx_perpetual_derivative - INFO - Created ordertype.limit tradetype.buy order buy-ETH-USD-1618196677001900 for 0.010 ETH-USD.
2021-04-12 11:04:38,309 - 3701 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1618196678000, "type": "OrderType.LIMIT", "trading_pair": "ETH-USD", "amount": "0.010", "price": "2117.3", "order_id": "buy-ETH-USD-1618196677001900", "exchange_order_id": null, "leverage": 1, "position": "NILL", "event_name": "BuyOrderCreatedEvent", "event_source": "dydx_perpetual"}
2021-04-12 11:04:38,314 - 3701 - hummingbot.connector.derivative.dydx_perpetual.dydx_perpetual_derivative - INFO - Created ordertype.limit tradetype.sell order sell-ETH-USD-1618196677002330 for 0.010 ETH-USD.
2021-04-12 11:04:38,327 - 3701 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1618196678000, "type": "OrderType.LIMIT", "trading_pair": "ETH-USD", "amount": "0.010", "price": "2138.9", "order_id": "sell-ETH-USD-1618196677002330", "exchange_order_id": null, "leverage": 1, "position": "NILL", "event_name": "SellOrderCreatedEvent", "event_source": "dydx_perpetual"}
2021-04-12 11:04:58,846 - 3701 - hummingbot.strategy.perpetual_market_making.perpetual_market_making - INFO - (ETH-USD) Maker sell order of 0.01 ETH filled. [clock=2021-04-12 03:04:58+00:00]
2021-04-12 11:04:58,916 - 3701 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1618196698.0, "order_id": "sell-ETH-USD-1618196677002330", "trading_pair": "ETH-USD", "trade_type": "TradeType.SELL", "order_type": "OrderType.LIMIT", "price": "2138.9", "amount": "0.01", "trade_fee": ["0", [["USD", "0.010694"]]], "exchange_trade_id": "sell-ETH-USD-1618196677002330", "leverage": 3, "position": "OPEN", "event_name": "OrderFilledEvent", "event_source": "dydx_perpetual"}
2021-04-12 11:04:58,917 - 3701 - hummingbot.connector.derivative.dydx_perpetual.dydx_perpetual_derivative - INFO - The market sell order sell-ETH-USD-1618196677002330 has completed according to user stream.
2021-04-12 11:04:58,919 - 3701 - hummingbot.strategy.perpetual_market_making.perpetual_market_making - INFO - (ETH-USD) Maker sell order sell-ETH-USD-1618196677002330 (0.010 ETH @ 2138.9 USD) has been completely filled. [clock=2021-04-12 03:04:58+00:00]
2021-04-12 11:04:58,931 - 3701 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1618196698.0, "order_id": "sell-ETH-USD-1618196677002330", "base_asset": "ETH", "quote_asset": "USD", "fee_asset": "USD", "base_asset_amount": "0.01", "quote_asset_amount": "21.389", "fee_amount": "0.228733966", "order_type": "OrderType.LIMIT", "exchange_order_id": null, "event_name": "SellOrderCompletedEvent", "event_source": "dydx_perpetual"}
2021-04-12 11:14:39,364 - 3701 - hummingbot.connector.derivative.dydx_perpetual.dydx_perpetual_derivative - INFO - Successfully cancelled order buy-ETH-USD-1618196677001900
2021-04-12 11:14:39,379 - 3701 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1618197279.0, "order_id": "buy-ETH-USD-1618196677001900", "exchange_order_id": null, "event_name": "OrderCancelledEvent", "event_source": "dydx_perpetual"}
```

Logs: [logs_dydx_perp.log](https://github.com/CoinAlpha/hummingbot/files/6294470/logs_dydx_perp.log)
Config: [dydx_perp.txt](https://github.com/CoinAlpha/hummingbot/files/6294474/dydx_perp.txt)

**Comment by RC-13** [comment]
Adding additional issues encountered while testing the connector. The test is done by running a perpetual market making strategy using profit-taking management
- Incorrect entry price once a position is open, if you check the image posted by @rxlxrxsx and my screenshot, we have the same entry price that doesn't change.
dydx-perp-entry-price

- Position doesn't get closed if the exit price is filled
dydx-perp-position-after-exit-price
dydx-exchange-position
dydx-exchange-fills

- no new order is created when the exit price is filled or if the profit-taking exit order expires (Good til time). Related to the issue above.

- balance seems to be added immediately assuming if the position is closed, I think this will cause confusion to users since they can manually close the position.
dydx-perp-balance
image

**Review by vic-en — CHANGES_REQUESTED** [review]

**Comment by rxlxrxsx** [comment]
After an order is filled to open a position, the bot will proceed to create orders like there's no position currently in place. The bot also doesn't recognize the open position once an order is filled upon checking on `status`. This happens both on `profit_taking` and `trailing_stop` position management.

Profit_taking
Screen Shot 2021-04-15 at 9 11 29 AM
Screen Shot 2021-04-15 at 9 11 50 AM

Trailing_stop
Screen Shot 2021-04-15 at 10 39 56 AM
Screen Shot 2021-04-15 at 10 40 25 AM

Logs and config: [dydx.zip](https://github.com/CoinAlpha/hummingbot/files/6314873/dydx.zip)

**Comment by trkoneill18** [comment]
rxlxrxsx, is this invariably the behavior that you see? I have failed to reproduce the problem. What configurations are you using for, say, profit taking?

**Comment by rxlxrxsx** [comment]
@trkoneill18 yes that's correct. Here is the config file with position management set to profit taking, you can also check the zip file I attached for logs and config file which is currently set to trailing stop as position management.

[dydx_perp.txt](https://github.com/CoinAlpha/hummingbot/files/6319113/dydx_perp.txt)

**Comment by trkoneill18** [comment]
I have just tested with those configurations and saw the proper behavior. I will check to see whether I missed anything on the last commit.

File: hummingbot/connector/derivative/dydx_perpetual/dydx_perpetual_derivative.py [hunk 1]
```diff
@@ -0,0 +1,1076 @@
...
+        await self.stop_network()
+        self._order_book_tracker.start()
+        if self._trading_required:
+            self._polling_update_task = safe_ensure_future(self._polling_update())
+            self._user_stream_tracker_task = safe_ensure_future(self._user_stream_tracker.start())
+            self._user_stream_event_listener_task = safe_ensure_future(self._user_stream_event_listener())
+
+    def _stop_network(self):
+        self._order_book_tracker.stop()
+        self._polling_update_task = None
```
Comment (vic-en) [inline hummingbot/connector/derivative/dydx_perpetual/dydx_perpetual_derivative.py:548]:
```suggestion
        self._order_book_tracker.stop()
        if self._polling_update_task is not None:
            self._polling_update_task.cancel()
            self._polling_update_task = None
```

**Review by vic-en — CHANGES_REQUESTED** [review]

**Comment by rxlxrxsx** [comment]
So far these are the issues currently happening on the latest commit which was also reported [above](https://github.com/CoinAlpha/hummingbot/pull/3130#issuecomment-817253819)

> Prompt for leverage still shows Binance Perpetual, leverage should also be 25x maximum leverage based on their website except for LINK-USD which has a maximum of 10x.

> After an order is filled to open a position, the bot will proceed to create orders like there's no position currently in place. The bot also doesn't recognize the open position once an order is filled upon checking on status. This happens both on profit_taking and trailing_stop position management.

File: hummingbot/connector/derivative/dydx_perpetual/dydx_perpetual_utils.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+from hummingbot.client.config.config_methods import using_exchange
+
+
+CENTRALIZED = True
+
+
+EXAMPLE_PAIR = "BTC-USD"
+
+
+DEFAULT_FEES = [0.02, 0.04]
```
Comment (vic-en) [inline hummingbot/connector/derivative/dydx_perpetual/dydx_perpetual_utils.py:11]:
```suggestion
DEFAULT_FEES = [0.05, 0.2]
```
According to the website, I see that's the maker-taker fee percent.

**Review by vic-en — CHANGES_REQUESTED** [review]

**Comment by trkoneill18** [comment]
> So far these are the issues currently happening on the latest commit which was also reported [above](https://github.com/CoinAlpha/hummingbot/pull/3130#issuecomment-817253819)
> 
> > Prompt for leverage still shows Binance Perpetual, leverage should also be 25x maximum leverage based on their website except for LINK-USD which has a maximum of 10x.
> 
> > After an order is filled to open a position, the bot will proceed to create orders like there's no position currently in place. The bot also doesn't recognize the open position once an order is filled upon checking on status. This happens both on profit_taking and trailing_stop position management.

Regarding the first issue, addressing this will require changes to the strategy configuration file. @vic-en would you like me to make the appropriate changes, or will this be handled by the hummingbot team?

As for the second, I have still been unable to reproduce this behavior. Has anyone else had the same issue?

**Comment by RC-13** [comment]
> As for the second, I have still been unable to reproduce this behavior. Has anyone else had the same issue?
Re-test the connector and still not seeing the open position, also can we request for the connector to be updated to the latest development version?

File: hummingbot/connector/derivative/dydx_perpetual/dydx_perpetual_utils.py (hunk 2 above)
Comment (dennisocana) [reply hummingbot/connector/derivative/dydx_perpetual/dydx_perpetual_utils.py]: Resolved in https://github.com/CoinAlpha/hummingbot/pull/3130/commits/c409bd3e9840b47e73eefa42d5264e7b7d6e87fd

File: hummingbot/connector/derivative/dydx_perpetual/dydx_perpetual_derivative.py (hunk 1 above)
Comment (dennisocana) [reply hummingbot/connector/derivative/dydx_perpetual/dydx_perpetual_derivative.py]: Resolved in https://github.com/CoinAlpha/hummingbot/pull/3130/commits/4c213f3d8146a51567ee6af4afcc92c3195ca119

File: hummingbot/VERSION [hunk 3]
```diff
@@ -1 +1 @@
-dev-0.38.0
+0.37.1
```
[the remark this answers is not in the document]
Comment (dennisocana) [reply hummingbot/VERSION:1]: Resolved in https://github.com/CoinAlpha/hummingbot/pull/3130/commits/16afe5fbb4514e0ae9bc258aa5b4a0140eac8d8e

File: hummingbot/client/config/config_helpers.py [hunk 4]
```diff
@@ -171,7 +171,7 @@ def get_erc20_token_addresses() -> Dict[str, List]:
     address_file_path = TOKEN_ADDRESSES_FILE_PATH
     token_list = {}
 
-    resp = requests.get(token_list_url, timeout=3)
+    resp = requests.get(token_list_url, timeout=1)
```
[the remark this answers is not in the document]
Comment (dennisocana) [reply hummingbot/client/config/config_helpers.py]: It seems this is the fix we implemented for the trading pair fetcher in the previous version. Let me know if I can resolve this comment.

https://github.com/CoinAlpha/hummingbot/blob/development/hummingbot/client/config/config_helpers.py#L174

**Comment by rxlxrxsx** [comment]
Same issue still persists on latest commit, no position is showed on the client after an order is filled to open a position. Both on profit_taking and trailing_stop as position_management

Trailing_stop
Screen Shot 2021-04-28 at 12 56 33 PM
Screen Shot 2021-04-28 at 12 56 54 PM

Profit_taking
Screen Shot 2021-04-28 at 1 07 15 PM
Screen Shot 2021-04-28 at 1 07 37 PM

**Comment by trkoneill18** [comment]
I have been able to reproduce the problem and have pushed up a fix for it. To reproduce the problem, I connected a new ethereum wallet to the exchange. One difference this appears to have caused was that I was unable to receive private data over the websocket, which in turn caused the connector to miss the position update. We are investigating potential causes and solutions for this.

**Comment by RC-13** [comment]
Position and entry price are now showing correctly however after the exit short price gets filled, the position remained in the bot however it's already closed in the exchange. This behavior prevents the bot from creating new orders.
image

Another issue is the fee shown in the performance. Two same fees is displayed instead of the total fee for the short and long filled positions
image
image

**Comment by trkoneill18** [comment]
I have pushed an update to the user stream data source that may fix the websocket problem, although it is unlikely that it would not have previously caused the same problem for me, having tested in - presumably - the same environment. 

I am curious about the steps you took to set up your client. I ran into some trouble setting up my own for the new wallet with which I have been testing recently. Namely, the api credentials I derived from https://docs.dydx.exchange/?json#recover-default-api-credentials were very slightly different from those that I discovered were actually stored in the client. It is possible that some such discrepancy is causing the problems with the websocket it appears that you are having. Since all of the restful api calls are done through the dydx client, and the websocket authentication makes explicit reference to the api keys that you store during connection, any such discrepancy would prohibit you from connecting to the private websocket feed while possibly allowing you to make api calls through the client.

**Comment by rxlxrxsx** [comment]
- dydx_perpetual is not showing on auto-complete when choosing a derivative connector. 
Screen Shot 2021-05-02 at 7 54 26 AM

Screen Shot 2021-05-02 at 7 55 01 AM

- The bot doesn't recognize the open position after restarting the strategy and proceeds to create orders like there's no position currently open.
Screen Shot 2021-05-02 at 8 42 52 AM

Screen Shot 2021-05-02 at 8 43 22 AM

Screen Shot 2021-05-02 at 8 43 42 AM

- Unable to place an close position order due to this error when `close_position_order_type` is set to MARKET.

Screen Shot 2021-05-02 at 11 51 20 AM

```
2021-05-02 11:38:09,714 - 4309 - hummingbot.connector.derivative.dydx_perpetual.dydx_perpetual_derivative - INFO - DydxApiError(status_code=400, response={'errors': [{'msg': 'Order with timeInForce: GTT, triggerPrice: undefined or trailingPercent: undefined will not be treated as MARKET'}]})
Traceback (most recent call last):
  File "/Users/reiniersalazar/hummingbot2/hummingbot/connector/derivative/dydx_perpetual/dydx_perpetual_derivative.py", line 348, in execute_order
    creation_response = await self.place_order(client_order_id, trading_pair, amount, order_side is TradeType.BUY, order_type, price, limit_fee, expiration)
  File "/Users/reiniersalazar/hummingbot2/hummingbot/connector/derivative/dydx_perpetual/dydx_perpetual_derivative.py", line 301, in place_order
    return await self.dydx_client.place_order(
  File "/Users/reiniersalazar/hummingbot2/hummingbot/connector/derivative/dydx_perpetual/dydx_perpetual_client_wrapper.py", line 56, in place_order
    return await f
  File "/Users/reiniersalazar/opt/anaconda3/envs/hummingbot/lib/python3.8/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/Users/reiniersalazar/opt/anaconda3/envs/hummingbot/lib/python3.8/site-packages/dydx3/modules/private.py", line 515, in create_order
    return self._post(
  File "/Users/reiniersalazar/opt/anaconda3/envs/hummingbot/lib/python3.8/site-packages/dydx3/modules/private.py", line 78, in _post
    return self._private_request(
  File "/Users/reiniersalazar/opt/anaconda3/envs/hummingbot/lib/python3.8/site-packages/dydx3/modules/private.py", line 64, in _private_request
    return request(
  File "/Users/reiniersalazar/opt/anaconda3/envs/hummingbot/lib/python3.8/site-packages/dydx3/helpers/requests.py", line 27, in request
    raise DydxApiError(response)
dydx3.errors.DydxApiError: DydxApiError(status_code=400, response={'errors': [{'msg': 'Order with timeInForce: GTT, triggerPrice: undefined or trailingPercent: undefined will not be treated as MARKET'}]})
```

Logs and config: [dydx_perpetual.zip](https://github.com/CoinAlpha/hummingbot/files/6410759/dydx_perpetual.zip)

File: hummingbot/VERSION (hunk 3 above)
[the remark this answers is not in the document]
Comment (keithbaum) [reply hummingbot/VERSION]: Resolved

**Comment by keithbaum** [comment]
@trkoneill18 @rxlxrxsx,
I pushed some minor changes to apply @vic-en changes, for funding_info error and for missing argument in MARKET orders.

@trkoneill18, could you please handle retrieving positions when connector starts?
Thanks

**Comment by keithbaum** [comment]
@trkoneill18 ,

I forced order_type to "LIMIT" in dydx_perpetual_derivative.py:Ln 300 because it was failing to submit proper market orders without parameters trailing_percent and trigger_price. Would you please check that and post a proper fix?
Thanks!

**Comment by RC-13** [comment]
Error with funding rate still persist but this time it shows a different message:
```python
2021-05-03 17:05:22,002 - 7198 - hummingbot.core.utils.async_utils - ERROR - Unhandled error in background task: 'decimal.Decimal' object is not subscriptable
Traceback (most recent call last):
  File "/Users/clint_g/Documents/Github-Repo/hb-dydx/hummingbot/core/utils/async_utils.py", line 9, in safe_wrapper
    return await c
  File "/Users/clint_g/Documents/Github-Repo/hb-dydx/hummingbot/strategy/spot_perpetual_arbitrage/spot_perpetual_arbitrage.py", line 160, in main
    if self.current_proposal.is_funding_payment_time():
  File "/Users/clint_g/Documents/Github-Repo/hb-dydx/hummingbot/strategy/spot_perpetual_arbitrage/arb_proposal.py", line 75, in is_funding_payment_time
    if self.timestamp > (funding_info["nextFundingTime"] - funding_payment_span[0]) and \
TypeError: 'decimal.Decimal' object is not subscriptable
```
image
[logs_conf_spot_perpetual_arbitrage_2.log](https://github.com/CoinAlpha/hummingbot/files/6413889/logs_conf_spot_perpetual_arbitrage_2.log)

**Comment by trkoneill18** [comment]
> @trkoneill18, could you please handle retrieving positions when connector starts?

Unfortunately, the last time I checked the relevant dydx endpoints were sending back dubious - or possibly deceptively labelled - data for the average entry price to a position. This necessitated the addition of a special dydx position class that tracked the status of the position. Things being as they are, the best that can be done on startup is an indication that a position exists and whether it is long or short; inaccurate price data could make this worse than not acknowledging the position at all. I will certainly update the connector upon confirmation that the necessary data is available from dydx, but it seems that this problem may require the addition of position data to tracking states in a fashion analogous to order data.

**Comment by rxlxrxsx** [comment]
- Filled trades are not getting notified when using `spot_perpetual_arbitrage` strategy

Screen Shot 2021-05-04 at 8 38 21 AM

- Orders price being executed on the logs are different from the orders price being placed on the exchange when using `spot_perpetual_arbitrage` strategy.

Screen Shot 2021-05-04 at 9 54 23 AM

First leg
Screen Shot 2021-05-04 at 10 00 06 AM

Screen Shot 2021-05-04 at 10 01 41 AM

Screen Shot 2021-05-04 at 10 02 24 AM

Second leg

Screen Shot 2021-05-04 at 10 00 43 AM

Screen Shot 2021-05-04 at 10 02 38 AM

Screen Shot 2021-05-04 at 10 02 47 AM

Logs and config: 
[spot_perp.zip](https://github.com/CoinAlpha/hummingbot/files/6418770/spot_perp.zip)

**Comment by keithbaum** [comment]
@rxlxrxsx @RC-13 ,
The message logged is mentioning the order original price, which ends up being slightly different to the executed price. I don't think that's an issue

**Comment by rxlxrxsx** [comment]
- There was one instance wherein the bot didn't follow the opening/closing of orders and position cycle:

1. Bot creates 1 buy 1 sell order to open position
2. Sell order got filled (3348.4) and cancelled the buy order (bot now has short position)
3. Bot creates a stop loss buy order (3351.7) to close short position
4. Stop loss buy order got filled (at this point, it should create orders to open a position again on the next cycle however)
Screen Shot 2021-05-05 at 10 00 40 AM

5. Bot creates a profit taking sell order (3358.4) after the stop loss buy order got filled for some reason
6. Profit taking sell order got filled
7. Bot creates another stop loss buy order (3361.8) after the profit taking sell order got filled
8. Stop loss buy order got filled
9. Bot creates another stop loss sell order again and eventually got cancelled, at this point I already stopped the bot

Screen Shot 2021-05-05 at 10 01 03 AM

Screen Shot 2021-05-05 at 6 44 57 PM

Please see attached log file for the full logging and to follow the behavior shown above
[dydx.zip](https://github.com/CoinAlpha/hummingbot/files/6426497/dydx.zip)

**Review by rxlxrxsx — DISMISSED** [review]
Tested and confirmed working by QA. No other issues found aside the position retrieval which @trkoneill18 is still looking into. QA will create a ticket for tracking regarding the position retrieval issue.

**Review by keithbaum — APPROVED** [review]
LGTM

**Review by vic-en — APPROVED** [review]
Approved to remove addressed but blocking requests.

**Comment by BaptisteGarcin** [comment]
@trkoneill18 do you think you could add the dydx testnet ?

Dydx testnet configuration:
 HTTP_HOST = 'https://api.stage.dydx.exchange'
 WS_HOST = 'wss://api.stage.dydx.exchange/v3/ws'

Dydx client should be instantiated with networkId: 3 
The Web3 config with the node endpoint should be set to a ropsten endpoint

**Pull request merged**